File: frontend/src/api/terminals.ts
import { apiClient } from './client';

export type TerminalSession = {
  id: string;
  owner: string;
  title: string;
  pid: number | null;
  rows: number;
  cols: number;
  status: string;
  output_cursor: number;
  replay_buffer_start_cursor: number;
  idle_timeout_minutes: number;
  created_at: string;
  last_active_at: string;
  exited_at: string | null;
  closed_at: string | null;
  close_reason: string | null;
};

export type TerminalHistory = {
  terminal_id: string;
  from_cursor: number;
  to_cursor: number;
  buffer_start_cursor: number;
  truncated: boolean;
  status: string;
  output: string;
};

function terminalPath(agentId: string, path = ''): string {
  if (agentId === 'local') {
    return `/api/terminals${path}`;
  }
  return `/api/agents/${encodeURIComponent(agentId)}/terminals${path}`;
}

export async function listTerminals(agentId: string, init?: RequestInit): Promise<TerminalSession[]> {
  const path = terminalPath(agentId);
  const response = init
    ? await apiClient.request<{ terminals: TerminalSession[] }>(path, init)
    : await apiClient.request<{ terminals: TerminalSession[] }>(path);
  return response.terminals;
}

export async function createTerminal(agentId: string, title = 'Terminal'): Promise<TerminalSession> {
  return apiClient.request<TerminalSession>(terminalPath(agentId), {
    method: 'POST',
    body: JSON.stringify({ title, rows: 24, cols: 80 }),
  });
}

export async function closeTerminal(agentId: string, id: string): Promise<TerminalSession> {
  return apiClient.request<TerminalSession>(terminalPath(agentId, `/${encodeURIComponent(id)}`), { method: 'DELETE' });
}

export async function resizeTerminal(agentId: string, id: string, rows: number, cols: number): Promise<void> {
  await apiClient.request<void>(terminalPath(agentId, `/${encodeURIComponent(id)}/resize`), {
    method: 'POST',
    body: JSON.stringify({ rows, cols }),
  });
}

export async function getTerminalHistory(agentId: string, id: string, cursor: number): Promise<TerminalHistory> {
  return apiClient.request<TerminalHistory>(
    terminalPath(agentId, `/${encodeURIComponent(id)}/history?cursor=${encodeURIComponent(String(cursor))}`),
  );
}

export async function createConnectToken(
  agentId: string,
  id: string,
): Promise<{ ticket: string; expires_in_seconds: number }> {
  return apiClient.request<{ ticket: string; expires_in_seconds: number }>(
    terminalPath(agentId, `/${encodeURIComponent(id)}/connect-token`),
    { method: 'POST' },
  );
}

File: backend/ic_env_guard/api/terminals.py
from typing import Annotated

from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field

from ic_env_guard.api.errors import ApiError
from ic_env_guard.auth.dependencies import AuthContext, require_auth
from ic_env_guard.terminal.manager import TerminalManager
from ic_env_guard.terminal.tickets import TerminalTicketManager

router = APIRouter(prefix="/api/terminals", tags=["terminals"])


class CreateTerminalRequest(BaseModel):
    title: str = Field(default="Terminal", max_length=80)
    rows: int = Field(default=24, ge=1)
    cols: int = Field(default=80, ge=1)
    cwd: str | None = None


class ResizeTerminalRequest(BaseModel):
    rows: int = Field(ge=1)
    cols: int = Field(ge=1)


class ConnectTokenResponse(BaseModel):
    ticket: str
    expires_in_seconds: int


def get_terminal_manager() -> TerminalManager:
    raise RuntimeError("TerminalManager dependency was not configured")


def get_ticket_manager() -> TerminalTicketManager:
    raise RuntimeError("TerminalTicketManager dependency was not configured")


def _terminal_or_404(manager: TerminalManager, terminal_id: str):
    try:
        return manager.get(terminal_id)
    except KeyError:
        raise ApiError(404, "not_found", "terminal session not found") from None


@router.get("")
def list_terminals(
    _: Annotated[AuthContext, Depends(require_auth)],
    manager: Annotated[TerminalManager, Depends(get_terminal_manager)],
) -> dict[str, list[dict[str, object]]]:
    return {"terminals": [session.to_dict() for session in manager.list()]}


@router.post("", status_code=201)
def create_terminal(
    payload: CreateTerminalRequest,
    auth: Annotated[AuthContext, Depends(require_auth)],
    manager: Annotated[TerminalManager, Depends(get_terminal_manager)],
) -> dict[str, object]:
    session = manager.create_terminal(
        title=payload.title,
        rows=payload.rows,
        cols=payload.cols,
        owner=auth.actor_id,
        cwd=payload.cwd,
    )
    return session.to_dict()


@router.get("/{terminal_id}")
def get_terminal(
    terminal_id: str,
    _: Annotated[AuthContext, Depends(require_auth)],
    manager: Annotated[TerminalManager, Depends(get_terminal_manager)],
) -> dict[str, object]:
    return _terminal_or_404(manager, terminal_id).to_dict()


@router.get("/{terminal_id}/history")
def get_terminal_history(
    terminal_id: str,
    cursor: int,
    _: Annotated[AuthContext, Depends(require_auth)],
    manager: Annotated[TerminalManager, Depends(get_terminal_manager)],
) -> dict[str, object]:
    _terminal_or_404(manager, terminal_id)
    history = manager.history(terminal_id, cursor)
    return {
        "terminal_id": history.terminal_id,
        "from_cursor": history.from_cursor,
        "to_cursor": history.to_cursor,
        "buffer_start_cursor": history.buffer_start_cursor,
        "truncated": history.truncated,
        "status": history.status,
        "output": history.output,
    }


@router.post("/{terminal_id}/connect-token", status_code=201)
def create_connect_token(
    terminal_id: str,
    _: Annotated[AuthContext, Depends(require_auth)],
    manager: Annotated[TerminalManager, Depends(get_terminal_manager)],
    tickets: Annotated[TerminalTicketManager, Depends(get_ticket_manager)],
) -> ConnectTokenResponse:
    _terminal_or_404(manager, terminal_id)
    ticket = tickets.issue(terminal_id)
    return ConnectTokenResponse(ticket=ticket.ticket, expires_in_seconds=tickets.ttl_seconds)


@router.post("/{terminal_id}/resize", status_code=204)
def resize_terminal(
    terminal_id: str,
    payload: ResizeTerminalRequest,
    _: Annotated[AuthContext, Depends(require_auth)],
    manager: Annotated[TerminalManager, Depends(get_terminal_manager)],
) -> None:
    _terminal_or_404(manager, terminal_id)
    manager.resize(terminal_id, payload.rows, payload.cols)


@router.delete("/{terminal_id}", status_code=202)
def close_terminal(
    terminal_id: str,
    _: Annotated[AuthContext, Depends(require_auth)],
    manager: Annotated[TerminalManager, Depends(get_terminal_manager)],
) -> dict[str, object]:
    _terminal_or_404(manager, terminal_id)
    return manager.close(terminal_id).to_dict()
